### PR TITLE
Transition GPUArrays to KernelAbstractions

### PR DESCRIPTION
**Pull request by vchuravy**
Depends on https://github.com/JuliaGPU/KernelAbstractions.jl/pull/317

Can happen last as long as we decide where backend lives.

### CODE REVIEW

File: src/device/execution.jl [hunk 1]
```diff
@@ -1,110 +1,8 @@
 # kernel execution
 
-export AbstractGPUBackend, AbstractKernelContext, gpu_call
-
-abstract type AbstractGPUBackend end
-
-abstract type AbstractKernelContext end
```
Comment (leios) [inline src/device/execution.jl:7]: Is the intention to remove `AbstractKernelContext` completely?

File: src/device/execution.jl [hunk 2]
```diff
@@ -1,110 +1,8 @@
...
-    else
-        (; threads, blocks, elements_per_thread=1)
-    end
-end
-
-gpu_call(backend::AbstractGPUBackend, kernel, args, threads::Int, blocks::Int; kwargs...) =
-    error("Not implemented") # COV_EXCL_LINE
+# TODO:
+# - Rename KA device to backend
+# - Who owns `AbstractGPUBackend`?
```
Comment (leios) [inline src/device/execution.jl:5]: Right now, `AbstractGPUBackend` is only used in the JLArrays lib, but can't we work towards deprecating that lib with KA's CPU capabilities?

If so, why do we need `AbstractGPUBackend`? Can't we just use KA's `Backend`?

File: src/host/random.jl [hunk 3]
```diff
@@ -84,29 +84,32 @@ function Random.seed!(rng::RNG, seed::Vector{UInt32})
 end
 
 function Random.rand!(rng::RNG, A::AnyGPUArray{T}) where T <: Number
-    gpu_call(A, rng.state) do ctx, a, randstates
-        idx = linear_index(ctx)
-        idx > length(a) && return
+    @kernel rand!(a, randstate)
+        idx = @index(Global, Linear)
         @inbounds a[idx] = gpu_rand(T, ctx, randstates)
```
Comment (leios) [inline src/host/random.jl:89]: Is this the only way people will be using `gpu_rand()`? Or will they be calling the functions on lines {33, 40, 58...}?

If so, we can just pass `idx` as `threadid` into the above functions, right?

**Review by leios — COMMENTED** [review]
Working on this PR (finally). Can we also just remove `indexing.jl` and `synchronization.jl` for KA's utilities?

File: src/device/execution.jl (hunk 1 above)
Comment (vchuravy) [reply src/device/execution.jl:7]: Yeah I think so. It shouldn't be needed anymore.

File: src/device/execution.jl (hunk 2 above)
Comment (vchuravy) [reply src/device/execution.jl:5]: I think we need a JLArrays implementation using KA first.

File: src/host/random.jl (hunk 3 above)
Comment (maleadt) [reply src/host/random.jl:89]: The only supported API are the Random stdlib functions; calling `gpu_rand` explicitly is not supported, so feel free to refactor.

**Comment by leios** [comment]
Great! I'm working on this for like an hour or two a day, so expect a PR next week? #517 is kinda in the right direction, but I have another branch that is rebased up to master so I can test ROC / CuArrays at least.

**Comment by christiangnrd** [comment]
Can this be closed?

**Pull request closed without merging**